File: leetcode/0668-kth-smallest-number-in-multiplication-table/solution.cpp
#include <algorithm>
using namespace std;

class Solution {
public:
    int findKthNumber(int m, int n, int k) {
        int lo = 1, hi = m * n;

        while (lo < hi) {
            int mid = lo + (hi - lo) / 2;
            if (countLessEqual(m, n, mid) >= k) {
                hi = mid;
            } else {
                lo = mid + 1;
            }
        }
        return lo;
    }

private:
    long long countLessEqual(int m, int n, int x) {
        long long total = 0;
        for (int i = 1; i <= m; i++) {
            total += min(x / i, n);
        }
        return total;
    }
};

File: leetcode/0668-kth-smallest-number-in-multiplication-table/solution.py
class Solution:
    def findKthNumber(self, m: int, n: int, k: int) -> int:
        def count_le(x: int) -> int:
            total = 0
            for i in range(1, m + 1):
                total += min(x // i, n)
            return total

        lo, hi = 1, m * n
        while lo < hi:
            mid = (lo + hi) // 2
            if count_le(mid) >= k:
                hi = mid
            else:
                lo = mid + 1
        return lo
